storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,811 @@
+    //run crawl, and do not reduce() the result, since reduce() calls us in turn
+    Fsm.crawl(resAlphabet, resInitials, isFinal, follow)
+  }
+
+  /**
+    * Return a finite state machine which will accept any string NOT
+    * accepted by self, and will not accept any string accepted by self.
+    * This is more complicated if there are missing transitions, because the
+    * missing "dead" state must now be reified.
+    */
+  def EverythingBut: Fsm = {
+    def resInitial = initialState :: Nil
+
+    def follow(currentState: List[Int], currentSymbol: Char): Option[List[Int]] = {
+      var next = List[Int]()
+      if (currentState.nonEmpty) {
+        val transition = transitions.get(currentState.head)
+        if (transition.isDefined) {
+          val nextState = transition.get.get(currentSymbol)
+          if (nextState.isDefined) {
+            next = nextState.get :: Nil
+          }
+        }
+      }
+      Some(next)
+    }
+
+    //state is final unless the original was
+    def isFinal(statesSet: List[Int]): Boolean =
+      !(statesSet.nonEmpty && finalStates.contains(statesSet.head))
+
+    Fsm.crawl(alphabet, resInitial, isFinal, follow).Reduced
+  }
+
+  /**
+    * Generate strings (lists of symbols) that this FSM accepts. Since there may
+    * be infinitely many of these we use a Stream[String] instead of constructing a
+    * static list. Strings will be sorted in order of length and then lexically.
+    * This procedure uses arbitrary amounts of memory but is very fast. There
+    * may be more efficient ways to do this, that I haven't investigated yet.
+    * You can use this in list comprehensions.
+    */
+  def Strings: Stream[String] = {
+    // Many FSMs have "dead states". Once you reach a dead state, you can no
+    // longer reach a final state. Since many strings may end up here, it's
+    // advantageous to constrain our search to live states only.
+    val livestates = states.filter(state => IsLive(state))
+
+    // We store a list of tuples. Each tuple consists of an input string and the
+    // state that this input string leads to. This means we don't have to run the
+    // state machine from the very beginning every time we want to check a new
+    // string.
+    val pending = Queue[(String, Int)]()
+    val pendingFinals = Queue[String]()
+    var needInitial = true
+
+    def nextStr: Option[String] =
+      if (pendingFinals.nonEmpty) {
+        Some(pendingFinals.dequeue())
+      } else {
+        //Initial entry (or possibly not, in which case this is a short one)
+        if (needInitial) {
+          needInitial = false
+          val cstate = this.initialState
+          if (livestates.contains(cstate)) {
+            pending += "" -> cstate
+            if (finalStates.contains(cstate)) {
+              pendingFinals += ""
+            }
+          }
+        }
+
+        while (pendingFinals.isEmpty && pending.nonEmpty) {
```

Comment:
Can we use recursion instead?

Reply:
We can, however we still need two queues and therefore will hardly win anything, just make algorithm less readable